// match-sort/lib/classes/MatchSortValues.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MatchSortIndex = exports.MatchSortValues = void 0;
var MatchSortValues = /** @class */ (function () {
    function MatchSortValues(sortFn, list) {
        this.lastSearch = '';
        this.searchFunction = sortFn;
        this.values = list;
        this.currentResult = list;
    }
    MatchSortValues.prototype.run = function (search) {
        var list = search.startsWith(this.lastSearch) ? this.currentResult : this.values;
        this.currentResult = this.searchFunction(search, list);
        this.lastSearch = search;
        return this.currentResult;
    };
    return MatchSortValues;
}());
exports.MatchSortValues = MatchSortValues;
var MatchSortIndex = /** @class */ (function () {
    function MatchSortIndex(searchFunction) {
        this.index = new Map();
        this.searchFunction = searchFunction;
    }
    MatchSortIndex.prototype.run = function (values, search) {
        return this.get(values).run(search);
    };
    MatchSortIndex.prototype.get = function (values) {
        if (!this.index.has(values))
            this.set(values);
        return this.index.get(values);
    };
    MatchSortIndex.prototype.set = function (values) {
        this.index.set(values, new MatchSortValues(this.searchFunction, values));
    };
    return MatchSortIndex;
}());
exports.MatchSortIndex = MatchSortIndex;
