// match-sort/lib/classes/SearchListIndex.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SearchListIndex = void 0;
var SearchListIndex = /** @class */ (function () {
    function SearchListIndex(searchFunction) {
        this.index = new Map();
        this.searchFunction = searchFunction;
    }
    SearchListIndex.prototype.run = function (values, search) {
        return this.get(values).run(search);
    };
    SearchListIndex.prototype.get = function (values) {
        if (!this.index.has(values))
            this.set(values);
        return this.index.get(values);
    };
    SearchListIndex.prototype.set = function (values) {
        this.index.set(values, new SearchListIndexInstance(this.searchFunction, values));
    };
    return SearchListIndex;
}());
exports.SearchListIndex = SearchListIndex;
var SearchListIndexInstance = /** @class */ (function () {
    function SearchListIndexInstance(sortFn, list) {
        this.lastSearch = '';
        this.searchFunction = sortFn;
        this.values = list;
        this.currentResult = list;
    }
    SearchListIndexInstance.prototype.run = function (search) {
        var list = search.startsWith(this.lastSearch) ? this.currentResult : this.values;
        if (search === this.lastSearch)
            return this.currentResult;
        this.currentResult = this.searchFunction(search, list);
        this.lastSearch = search;
        return this.currentResult;
    };
    return SearchListIndexInstance;
}());
